### DISCUSSION TITLE
Simplifying geometries while preserving coverage topology

### DISCUSSION THREAD

**Discussion by nshiab** [duckdb/duckdb-spatial#293] [Q&A] [answered]
Hi!

Are there any plans to add something like this?
https://postgis.net/docs/ST_CoverageSimplify.html

If I am not mistaken, all functions in the library right now simplify the objects one by one. So, if you simplify something like the world countries' borders, the borders won't be contiguous anymore.

Thank you for your amazing work. :)

**Comment by Maxxen** [comment]
Hi! Thanks for the suggestion! This definitely sounds like a good function to implement, I'll look into it!

**Reply by nshiab** [reply]
Thank you! 🥳

**Comment by nshiab** [comment]
Hi @Maxxen!

I was wondering if you still plan to add this function at some point.

Cheers

**Reply by Maxxen** [reply]
Hi! Sorry, I lost track of this. This shouldn't be very hard to add at all, especially now that we've refactored the whole GEOS integration. I'll try to get it in before the next release. 1.2.1 is scheduled for 2025-03-05.

**Reply by nshiab** [reply]
Amazing! Thank you!

**Reply by nshiab** [reply]
Hi @Maxxen ! Just curious: were you able to squeeze it in?

**Reply by Maxxen** [reply]
On the way! https://github.com/duckdb/duckdb-spatial/pull/532

**Reply by nshiab** [reply]
Amazing!!

**Comment by nshiab** [comment]
Very sorry to bother you again @Maxxen...

But I am trying to understand how to use the new functions `ST_CoverageSimplify` and `ST_CoverageSimplify_Agg`.

How would you use them to simplify a file, such as this one, for example?
https://github.com/nshiab/simple-data-analysis/blob/main/test/geodata/files/CanadianProvincesAndTerritories.json

Thank you! 🙏

**Reply by nshiab** [reply]
I think I got it right. I passed the function documentation to Gemini, which has been super helpful.

Here's how I made it work in my queries.

```sql
CREATE OR REPLACE TABLE "table1" AS SELECT * REPLACE(ST_CoverageSimplify(ARRAY_AGG("geom"), 0.5) AS "geom") FROM "table1" GROUP BY ALL;
```

Thank you for implementing it, @Maxxen !

**Reply by Maxxen** [reply, marked as answer]
No problem! This is basically equivalent to using the aggregate version though:
```
SELECT st_coveragesimplify_agg(geom, 0.5) from st_read('https://raw.githubusercontent.com/nshiab/simple-data-analysis/refs/heads/main/test/geodata/files/CanadianProvincesAndTerritories.json');
```
As far as what is faster... not sure. This avoids creating a LIST (with ARRAY_AGG) to hold the input geometries, so should be more efficient in that sense, at the same time it might keep more untracked memory alive longer.

**Reply by nshiab** [reply]
Ahhhh! Thank you!
